# Hoja de Trabajo \# 4

---


por Josué Obregón <br>
DS6011 - Feature Engineering <br>
UVG Masters - Escuela de Negocios<br>


## Objetivos

El objetivo de esta hoja de trabajo  es presentar al estudiante diferentes técnicas de codificación de variables categóricas. Éstas técnicas incluyen codificadores clásicos, de contraste y codificadores supervisados o bayesianos.

También se busca que el estudiante practique la utilización de éstas técnicas con las librerías disponibles en el lenguaje Python.


## Importación de librerías y carga de los datos a varios pandas [DataFrames](https://pandas.pydata.org/pandas-docs/version/1.1.5/reference/frame.html)




In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

In [2]:
!mkdir data

In [3]:
import gdown

urls = ['https://drive.google.com/uc?export=download&id=16AGQw1nM9NYILv2aSZaSNSn9jBPByWPq', # okc_train  https://drive.google.com/file/d/16AGQw1nM9NYILv2aSZaSNSn9jBPByWPq/view?usp=sharing
        ]
outputs = ['okc_train.csv']
for url,output in zip(urls,outputs):
  gdown.download(url, f'data/{output}', quiet=False)

Downloading...
From: https://drive.google.com/uc?export=download&id=16AGQw1nM9NYILv2aSZaSNSn9jBPByWPq
To: /content/data/okc_train.csv
12.3MB [00:00, 169MB/s]


In [4]:
df = pd.read_csv('data/okc_train.csv',index_col=0)

In [5]:
df.head()

age               diet      drinks  ... essay_length  Class  profile
1    22  strictly_anything    socially  ...     3.420945  other        1
4    23         vegetarian    socially  ...     2.671173  other        3
6    29    mostly_anything    socially  ...     3.391112   stem        5
8    31    mostly_anything    socially  ...     3.090611  other        6
10   37    mostly_anything  not_at_all  ...     3.382377  other        7

[5 rows x 38 columns]

In [6]:
df['drinks'].value_counts()

socially          27521
rarely             3981
often              3413
not_at_all         2157
drinks_missing     1208
very_often          319
desperately         210
Name: drinks, dtype: int64

In [7]:
df['status'].value_counts()

single            36014
seeing_someone     1354
available          1234
married             204
unknown               3
Name: status, dtype: int64

Pequeño dataset de prueba para algunas explicaciones y para las figuras en las diapositivas


In [8]:
df_test = pd.DataFrame({
... 'City': ['SF', 'SF', 'SF', 'NYC', 'NYC', 'NYC',
... 'Seattle', 'Seattle', 'Seattle'],
... 'Rent': [3999, 4000, 4001, 3499, 3500, 3501, 2499, 2500, 2501]
... })

In [9]:
df_test

City  Rent
0       SF  3999
1       SF  4000
2       SF  4001
3      NYC  3499
4      NYC  3500
5      NYC  3501
6  Seattle  2499
7  Seattle  2500
8  Seattle  2501

Para esta hoja de trabajo vamos a utilizar la librería [category_encoders](http://contrib.scikit-learn.org/category_encoders/index.html), la cual es compatible con scikit-learn.

In [10]:
!pip install category_encoders

     |████████████████████████████████| 81kB 3.8MB/s 


In [11]:
col = 'drinks'

# Unsupervised Encoders

## Classic Encoders

### One-Hot Encoder

In [12]:
from category_encoders import OneHotEncoder

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning:

pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.



In [13]:
onehot_enc=OneHotEncoder(use_cat_names=True)

In [14]:
onehot_enc.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks_socially  ...  drinks_desperately
1                    1  ...                   0
4                    1  ...                   0
6                    1  ...                   0
8                    1  ...                   0
10                   0  ...                   0
...                ...  ...                 ...
59940                1  ...                   0
59943                0  ...                   0
59944                0  ...                   0
59945                1  ...                   0
59946                1  ...                   0

[38809 rows x 7 columns]

In [15]:
onehot_enc.mapping[0]['mapping']

drinks_socially  drinks_not_at_all  ...  drinks_very_often  drinks_desperately
 1                1                  0  ...                  0                   0
 2                0                  1  ...                  0                   0
 3                0                  0  ...                  0                   0
 4                0                  0  ...                  0                   0
 5                0                  0  ...                  0                   0
 6                0                  0  ...                  1                   0
 7                0                  0  ...                  0                   1
-1                0                  0  ...                  0                   0
-2                0                  0  ...                  0                   0

[9 rows x 7 columns]

### Binary Encoder

In [16]:
from category_encoders import BinaryEncoder

In [17]:
binary_enc=BinaryEncoder()

In [18]:
binary_enc.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks_0  drinks_1  drinks_2  drinks_3
1             0         0         0         1
4             0         0         0         1
6             0         0         0         1
8             0         0         0         1
10            0         0         1         0
...         ...       ...       ...       ...
59940         0         0         0         1
59943         0         1         0         0
59944         0         0         1         0
59945         0         0         0         1
59946         0         0         0         1

[38809 rows x 4 columns]

In [19]:
binary_enc.base_n_encoder.mapping[0]['mapping']

drinks_0  drinks_1  drinks_2  drinks_3
 1         0         0         0         1
 2         0         0         1         0
 3         0         0         1         1
 4         0         1         0         0
 5         0         1         0         1
 6         0         1         1         0
 7         0         1         1         1
-1         0         0         0         0
-2         0         0         0         0

### Frequency or Count Encoder

In [20]:
from category_encoders import CountEncoder

In [21]:
count_enc=CountEncoder()

In [22]:
count_enc.fit_transform(df[[col]])

drinks
1       27521
4       27521
6       27521
8       27521
10       2157
...       ...
59940   27521
59943    3413
59944    2157
59945   27521
59946   27521

[38809 rows x 1 columns]

In [23]:
count_enc.mapping

{'drinks': socially          27521
 rarely             3981
 often              3413
 not_at_all         2157
 drinks_missing     1208
 very_often          319
 desperately         210
 Name: drinks, dtype: int64}

Intentemos utilizando las funciones de combinacion de grupos

In [24]:
countmin_enc=CountEncoder(combine_min_nan_groups=True,min_group_size=500,min_group_name='otros')

In [25]:
countmin_enc.fit_transform(df[[col]])

drinks
1       27521
4       27521
6       27521
8       27521
10       2157
...       ...
59940   27521
59943    3413
59944    2157
59945   27521
59946   27521

[38809 rows x 1 columns]

In [26]:
countmin_enc.mapping

{'drinks': socially          27521
 rarely             3981
 often              3413
 not_at_all         2157
 drinks_missing     1208
 otros               529
 Name: drinks, dtype: int64}

### Ordinal Encoder

In [27]:
from category_encoders import OrdinalEncoder

In [28]:
ordinal_enc=OrdinalEncoder()

In [29]:
ordinal_enc.fit_transform(df[[col]])

drinks
1           1
4           1
6           1
8           1
10          2
...       ...
59940       1
59943       4
59944       2
59945       1
59946       1

[38809 rows x 1 columns]

In [30]:
ordinal_enc.mapping

[{'col': 'drinks', 'data_type': dtype('O'), 'mapping': socially          1
  not_at_all        2
  rarely            3
  often             4
  drinks_missing    5
  very_often        6
  desperately       7
  NaN              -2
  dtype: int64}]

In [31]:
drink_dict={None:0,'not_at_all':1,'rarely':2,'socially':3,'often':4,'very_often':5,'desperately':6,'drinks_missing':-1}
col_drink_dict={'col':'drinks','mapping':drink_dict}

In [32]:
ord_enc=OrdinalEncoder(mapping=[col_drink_dict])

In [33]:
ord_enc.fit_transform(df[col],)

drinks
1           3
4           3
6           3
8           3
10          1
...       ...
59940       3
59943       4
59944       1
59945       3
59946       3

[38809 rows x 1 columns]

###Feature Hashing

In [34]:
from sklearn.feature_extraction import FeatureHasher

In [35]:
col='where_town'

In [36]:
hash_enc=FeatureHasher(n_features=8,input_type='string',alternate_sign=True)

In [37]:
hashed_features=hash_enc.fit_transform(df[col])

In [38]:
hashed_features.toarray()

array([[ 1., -1.,  2., ..., -2.,  0., -3.],
       [ 0.,  1.,  0., ..., -1.,  1.,  2.],
       [ 0., -1.,  2., ..., -1.,  0., -2.],
       ...,
       [ 1., -1.,  2., ..., -2.,  0., -3.],
       [ 0., -1.,  2., ..., -1.,  0., -2.],
       [ 0., -1.,  2., ..., -1.,  0., -2.]])

In [39]:
np.unique(np.array(hashed_features.todense()),axis=0).shape

(50, 8)

In [40]:
from sys import getsizeof

In [41]:
print('Panda: ',getsizeof(df[col]))
print('Hashed: ',getsizeof(hashed_features))

Panda:  2944494
Hashed:  64


## Contrast Encoders

### Diferencia entre One-hot Encoding y Dummy Coding

In [42]:
df_test

City  Rent
0       SF  3999
1       SF  4000
2       SF  4001
3      NYC  3499
4      NYC  3500
5      NYC  3501
6  Seattle  2499
7  Seattle  2500
8  Seattle  2501

In [43]:
print('Global: ',df_test['Rent'].mean())
print(df_test.groupby('City').mean())

Global:  3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


In [44]:
one_hot_df=pd.get_dummies(df_test,prefix=['city'])

In [45]:
one_hot_df

Rent  city_NYC  city_SF  city_Seattle
0  3999         0        1             0
1  4000         0        1             0
2  4001         0        1             0
3  3499         1        0             0
4  3500         1        0             0
5  3501         1        0             0
6  2499         0        0             1
7  2500         0        0             1
8  2501         0        0             1

In [46]:
one_hot_df[['city_NYC','city_SF','city_Seattle']].drop_duplicates()

city_NYC  city_SF  city_Seattle
0         0        1             0
3         1        0             0
6         0        0             1

Con la codificación one-hot, el término de intersección (intercepto) representa la media global de la variable objetivo: 'Renta', y cada uno de los coeficientes lineales representa que tanto difiere la renta media de esa ciudad respecto a la media global.

In [47]:
from sklearn import linear_model

In [48]:
model=linear_model.LinearRegression()
model.fit(one_hot_df[['city_NYC','city_SF','city_Seattle']],one_hot_df['Rent'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [49]:
print(f'Coefficients: {model.coef_}')
print(f'Incercept: {model.intercept_}')

Coefficients: [ 166.66666667  666.66666667 -833.33333333]
Incercept: 3333.3333333333335


Ahora entrenemos el mismo modelo con dummy coding

In [50]:
dummy_df=pd.get_dummies(df_test,prefix=['city'],drop_first=True)
dummy_df

Rent  city_SF  city_Seattle
0  3999        1             0
1  4000        1             0
2  4001        1             0
3  3499        0             0
4  3500        0             0
5  3501        0             0
6  2499        0             1
7  2500        0             1
8  2501        0             1

In [51]:
model.fit(dummy_df[['city_SF','city_Seattle']],dummy_df['Rent'])
print(f'Coefficients: {model.coef_}')
print(f'Incercept: {model.intercept_}')

Coefficients: [  500. -1000.]
Incercept: 3500.0


Con codificación dummy, el coeficiente de sesgo (intercepto) representa el valor medio de la variable $y$ para la categoría de referencia, que en el ejemplo es la ciudad NYC. El coeficiente para la *i*-ésima característica es igual a la diferencia entre el valor medio de la respuesta para la i-ésima categoría y la media de la categoría de referencia.

### Dummy Coding

In [52]:
from sklearn.preprocessing import OneHotEncoder as OneHotEncoder_sk

In [53]:
col='drinks'

In [54]:
dummy_enc=OneHotEncoder_sk(drop='first',sparse=False)

In [55]:
dummy_enc.fit_transform(df[[col]])

array([[0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1., 0.],
       ...,
       [0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1., 0.]])

In [56]:
dummy_enc.categories_

[array(['desperately', 'drinks_missing', 'not_at_all', 'often', 'rarely',
        'socially', 'very_often'], dtype=object)]

In [57]:
dummy_df2=pd.get_dummies(df[col],prefix='dr',drop_first=True) #dummy_nabool
dummy_df2.head()

dr_drinks_missing  dr_not_at_all  ...  dr_socially  dr_very_often
1                   0              0  ...            1              0
4                   0              0  ...            1              0
6                   0              0  ...            1              0
8                   0              0  ...            1              0
10                  0              1  ...            0              0

[5 rows x 6 columns]

In [58]:
dummy_df2.drop_duplicates()

dr_drinks_missing  dr_not_at_all  ...  dr_socially  dr_very_often
1                    0              0  ...            1              0
10                   0              1  ...            0              0
19                   0              0  ...            0              0
41                   0              0  ...            0              0
93                   1              0  ...            0              0
97                   0              0  ...            0              1
549                  0              0  ...            0              0

[7 rows x 6 columns]


### Sum (or Deviation) Coding

In [59]:
from category_encoders import SumEncoder

In [60]:
sum_enc=SumEncoder()

In [61]:
sum_enc.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  drinks_0  drinks_1  drinks_2  drinks_3  drinks_4  drinks_5
1              1       1.0       0.0       0.0       0.0       0.0       0.0
4              1       1.0       0.0       0.0       0.0       0.0       0.0
6              1       1.0       0.0       0.0       0.0       0.0       0.0
8              1       1.0       0.0       0.0       0.0       0.0       0.0
10             1       0.0       1.0       0.0       0.0       0.0       0.0
...          ...       ...       ...       ...       ...       ...       ...
59940          1       1.0       0.0       0.0       0.0       0.0       0.0
59943          1       0.0       0.0       0.0       1.0       0.0       0.0
59944          1       0.0       1.0       0.0       0.0       0.0       0.0
59945          1       1.0       0.0       0.0       0.0       0.0       0.0
59946          1       1.0       0.0       0.0       0.0       0.0       0.0

[38809 rows x 7 columns]

In [62]:
sum_enc.mapping[0]['mapping']

drinks_0  drinks_1  drinks_2  drinks_3  drinks_4  drinks_5
 1       1.0       0.0       0.0       0.0       0.0       0.0
 2       0.0       1.0       0.0       0.0       0.0       0.0
 3       0.0       0.0       1.0       0.0       0.0       0.0
 4       0.0       0.0       0.0       1.0       0.0       0.0
 5       0.0       0.0       0.0       0.0       1.0       0.0
 6       0.0       0.0       0.0       0.0       0.0       1.0
 7      -1.0      -1.0      -1.0      -1.0      -1.0      -1.0
-1       0.0       0.0       0.0       0.0       0.0       0.0
-2       0.0       0.0       0.0       0.0       0.0       0.0

Utilizando el ejemplo de prueba

In [63]:
sum_enc=SumEncoder()
sum_df=sum_enc.fit_transform(df_test['City'])
sum_df

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  City_0  City_1
0          1     1.0     0.0
1          1     1.0     0.0
2          1     1.0     0.0
3          1     0.0     1.0
4          1     0.0     1.0
5          1     0.0     1.0
6          1    -1.0    -1.0
7          1    -1.0    -1.0
8          1    -1.0    -1.0

In [64]:
sum_enc.ordinal_encoder.category_mapping[0]['mapping']

SF         1
NYC        2
Seattle    3
NaN       -2
dtype: int64

In [65]:
sum_df['Rent']=df_test['Rent']
sum_df

intercept  City_0  City_1  Rent
0          1     1.0     0.0  3999
1          1     1.0     0.0  4000
2          1     1.0     0.0  4001
3          1     0.0     1.0  3499
4          1     0.0     1.0  3500
5          1     0.0     1.0  3501
6          1    -1.0    -1.0  2499
7          1    -1.0    -1.0  2500
8          1    -1.0    -1.0  2501

In [66]:
model=linear_model.LinearRegression()
model.fit(sum_df[['City_0','City_1']],sum_df['Rent'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

La codificación de efectos (sum coding) es muy similar a la codificación dummy, pero da como resultado modelos de regresión lineal que son aún más simples de interpretar.

En el ejemplo, vemos que el término de intersección representa la media global del la variable respuesta, y los coeficientes individuales indican cuánto las medias de las categorías individuales difieren de la media global. (Esto se llama el efecto principal de la categoría o nivel, de ahí el nombre "codificación de efectos").

La codificación One-hot, encontró la misma intersección y coeficientes, pero en ese caso hay coeficientes lineales para cada ciudad. En la codificación de efecto, ningun feature representa la categoría de referencia, ***por lo que el efecto de la categoría de referencia debe calcularse por separado como la suma negativa de los coeficientes de todas las demás categorías.***

In [67]:
print(f'Coefficients:{model.coef_}')
print(f'Intercept:{model.intercept_}')
print(f'Negative sum of coefficents:{np.sum(model.coef_*-1)}')

Coefficients:[666.66666667 166.66666667]
Intercept:3333.3333333333335
Negative sum of coefficents:-833.3333333333337


In [68]:
print('Global mean: ',df_test['Rent'].mean())
print(df_test.groupby('City').mean())

Global mean:  3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


Restulados de la regresión lineal con One-hot encoding

Coefficients: $[ 166.66666667,   666.66666667,  -833.33333333]$

Intercept: $3333.3333333333335$

### Backward Difference Coding

In [69]:
from category_encoders import BackwardDifferenceEncoder

In [70]:
back_enc=BackwardDifferenceEncoder()

In [71]:
back_enc.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  drinks_0  drinks_1  drinks_2  drinks_3  drinks_4  drinks_5
1              1 -0.857143 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
4              1 -0.857143 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
6              1 -0.857143 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
8              1 -0.857143 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
10             1  0.142857 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
...          ...       ...       ...       ...       ...       ...       ...
59940          1 -0.857143 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
59943          1  0.142857  0.285714  0.428571 -0.428571 -0.285714 -0.142857
59944          1  0.142857 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
59945          1 -0.857143 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
59946          1 -0.857143 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857

[38809 rows x 7 columns]

In [72]:
back_enc.mapping[0]['mapping']

drinks_0  drinks_1  drinks_2  drinks_3  drinks_4  drinks_5
 1 -0.857143 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
 2  0.142857 -0.714286 -0.571429 -0.428571 -0.285714 -0.142857
 3  0.142857  0.285714 -0.571429 -0.428571 -0.285714 -0.142857
 4  0.142857  0.285714  0.428571 -0.428571 -0.285714 -0.142857
 5  0.142857  0.285714  0.428571  0.571429 -0.285714 -0.142857
 6  0.142857  0.285714  0.428571  0.571429  0.714286 -0.142857
 7  0.142857  0.285714  0.428571  0.571429  0.714286  0.857143
-1  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000
-2  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000

Utilizando el ejemplo de prueba

In [73]:
back_enc=BackwardDifferenceEncoder()
back_df=back_enc.fit_transform(df_test['City'])
back_df['Rent']=df_test['Rent']
back_df

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept    City_0    City_1  Rent
0          1 -0.666667 -0.333333  3999
1          1 -0.666667 -0.333333  4000
2          1 -0.666667 -0.333333  4001
3          1  0.333333 -0.333333  3499
4          1  0.333333 -0.333333  3500
5          1  0.333333 -0.333333  3501
6          1  0.333333  0.666667  2499
7          1  0.333333  0.666667  2500
8          1  0.333333  0.666667  2501

In [74]:
back_enc.mapping[0]['mapping']

City_0    City_1
 1 -0.666667 -0.333333
 2  0.333333 -0.333333
 3  0.333333  0.666667
-1  0.000000  0.000000
-2  0.000000  0.000000

In [75]:
back_enc.ordinal_encoder.category_mapping[0]['mapping']

SF         1
NYC        2
Seattle    3
NaN       -2
dtype: int64

In [76]:
model=linear_model.LinearRegression()
model.fit(back_df[['City_0','City_1']],back_df['Rent'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

La codificación de diferencia en reversa es útil para codificar variables ordinales.

En el ejemplo, vemos que el término de intersección representa la media global de la variable respuesta, y los coeficientes individuales indican cuánto las medias de las categorías individuales difieren de la media de la categoría inmediatamente anterior.

In [77]:
print(f'Coefficients:{model.coef_}')
print(f'Intercept:{model.intercept_}')
print(f'Negative sum of coefficents:{np.sum(model.coef_*-1)}')

Coefficients:[ -500. -1000.]
Intercept:3333.3333333333335
Negative sum of coefficents:1500.0000000000002


In [78]:
print('Global mean: ',df_test['Rent'].mean())
print(df_test.groupby('City').mean())

Global mean:  3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


### Helmert Coding

In [79]:
from category_encoders import HelmertEncoder

In [80]:
helm_enc=HelmertEncoder()

In [81]:
helm_enc.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  drinks_0  drinks_1  drinks_2  drinks_3  drinks_4  drinks_5
1              1      -1.0      -1.0      -1.0      -1.0      -1.0      -1.0
4              1      -1.0      -1.0      -1.0      -1.0      -1.0      -1.0
6              1      -1.0      -1.0      -1.0      -1.0      -1.0      -1.0
8              1      -1.0      -1.0      -1.0      -1.0      -1.0      -1.0
10             1       1.0      -1.0      -1.0      -1.0      -1.0      -1.0
...          ...       ...       ...       ...       ...       ...       ...
59940          1      -1.0      -1.0      -1.0      -1.0      -1.0      -1.0
59943          1       0.0       0.0       3.0      -1.0      -1.0      -1.0
59944          1       1.0      -1.0      -1.0      -1.0      -1.0      -1.0
59945          1      -1.0      -1.0      -1.0      -1.0      -1.0      -1.0
59946          1      -1.0      -1.0      -1.0      -1.0      -1.0      -1.0

[38809 rows x 7 columns]

In [82]:
helm_enc.mapping[0]['mapping']

drinks_0  drinks_1  drinks_2  drinks_3  drinks_4  drinks_5
 1      -1.0      -1.0      -1.0      -1.0      -1.0      -1.0
 2       1.0      -1.0      -1.0      -1.0      -1.0      -1.0
 3       0.0       2.0      -1.0      -1.0      -1.0      -1.0
 4       0.0       0.0       3.0      -1.0      -1.0      -1.0
 5       0.0       0.0       0.0       4.0      -1.0      -1.0
 6       0.0       0.0       0.0       0.0       5.0      -1.0
 7       0.0       0.0       0.0       0.0       0.0       6.0
-1       0.0       0.0       0.0       0.0       0.0       0.0
-2       0.0       0.0       0.0       0.0       0.0       0.0

Utilizando el ejemplo de prueba

In [83]:
helm_enc=HelmertEncoder()
helm_df=helm_enc.fit_transform(df_test['City'])
helm_df['Rent']=df_test['Rent']
helm_df

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  City_0  City_1  Rent
0          1    -1.0    -1.0  3999
1          1    -1.0    -1.0  4000
2          1    -1.0    -1.0  4001
3          1     1.0    -1.0  3499
4          1     1.0    -1.0  3500
5          1     1.0    -1.0  3501
6          1     0.0     2.0  2499
7          1     0.0     2.0  2500
8          1     0.0     2.0  2501

In [84]:
helm_enc.mapping[0]['mapping']

City_0  City_1
 1    -1.0    -1.0
 2     1.0    -1.0
 3     0.0     2.0
-1     0.0     0.0
-2     0.0     0.0

In [85]:
model=linear_model.LinearRegression()
model.fit(helm_df[['City_0','City_1']],helm_df['Rent'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [86]:
print(f'Coefficients:{model.coef_}')
print(f'Intercept:{model.intercept_}')
print(f'Negative sum of coefficents:{np.sum(model.coef_*-1)}')

Coefficients:[-250.         -416.66666667]
Intercept:3333.3333333333335
Negative sum of coefficents:666.666666666667


In [87]:
print('Global mean: ',df_test['Rent'].mean())
print(df_test.groupby('City').mean())

Global mean:  3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


#Supervised Encoders

In [88]:
col_cat='drinks'
col_num='essay_length'

In [89]:
from sklearn.preprocessing import LabelEncoder

In [90]:
label_enc=LabelEncoder()
df['Class_enc']=label_enc.fit_transform(df['Class'])

In [91]:
df.head()

age               diet      drinks  ...  Class profile  Class_enc
1    22  strictly_anything    socially  ...  other       1          0
4    23         vegetarian    socially  ...  other       3          0
6    29    mostly_anything    socially  ...   stem       5          1
8    31    mostly_anything    socially  ...  other       6          0
10   37    mostly_anything  not_at_all  ...  other       7          0

[5 rows x 39 columns]

## Target Encoder

In [92]:
from category_encoders import TargetEncoder

In [93]:
target_enc=TargetEncoder()
target_enc.fit_transform(df[col_cat],df['Class_enc'])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      0.189201
4      0.189201
6      0.189201
8      0.189201
10     0.147891
...         ...
59940  0.189201
59943  0.195429
59944  0.147891
59945  0.189201
59946  0.189201

[38809 rows x 1 columns]

In [94]:
target_enc.mapping

{'drinks': drinks
  1    0.189201
  2    0.147891
  3    0.170058
  4    0.195429
  5    0.179636
  6    0.137931
  7    0.171429
 -1    0.184674
 -2    0.184674
 dtype: float64}

In [95]:
target_enc=TargetEncoder()
target_enc.fit_transform(df[col_cat],df[col_num])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      3.136058
4      3.136058
6      3.136058
8      3.136058
10     3.145403
...         ...
59940  3.136058
59943  3.138155
59944  3.145403
59945  3.136058
59946  3.136058

[38809 rows x 1 columns]

In [96]:
target_enc.mapping

{'drinks': drinks
  1    3.136058
  2    3.145403
  3    3.164222
  4    3.138155
  5    2.963267
  6    2.915820
  7    3.072538
 -1    3.132119
 -2    3.132119
 dtype: float64}

## Leave-one-out Encoder

In [97]:
from category_encoders import LeaveOneOutEncoder

In [98]:
lou_enc=LeaveOneOutEncoder(sigma=0.05)
lou_enc.fit_transform(df[col_cat],df['Class_enc'])

drinks
1      0.191306
4      0.189456
6      0.192664
8      0.194475
10     0.142623
...         ...
59940  0.194674
59943  0.201259
59944  0.142417
59945  0.195184
59946  0.178406

[38809 rows x 1 columns]

In [99]:
lou_enc.mapping

{'drinks':                    sum  count
 desperately       36.0    210
 drinks_missing   217.0   1208
 not_at_all       319.0   2157
 often            667.0   3413
 rarely           677.0   3981
 socially        5207.0  27521
 very_often        44.0    319}

In [100]:
lou_enc=LeaveOneOutEncoder(sigma=0.05)
lou_enc.fit_transform(df[col_cat],df['essay_length'])

drinks
1      3.340353
4      3.114343
6      2.928397
8      3.104584
10     2.993493
...         ...
59940  3.219426
59943  3.175029
59944  3.287039
59945  3.370336
59946  3.312879

[38809 rows x 1 columns]

In [101]:
lou_enc.mapping

{'drinks':                          sum  count
 desperately       645.232955    210
 drinks_missing   3579.626046   1208
 not_at_all       6784.635061   2157
 often           10710.522351   3413
 rarely          12596.767259   3981
 socially        86307.461805  27521
 very_often        930.146539    319}

## M-estimate Encoder

In [ ]:
from category_encoders import MEstimateEncoder

In [ ]:
mest_enc=MEstimateEncoder(m=10)
mest_enc.fit_transform(df[col_cat],df['Class_enc'])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      0.189199
4      0.189199
6      0.189199
8      0.189199
10     0.148060
...         ...
59940  0.189199
59943  0.195398
59944  0.148060
59945  0.189199
59946  0.189199

[38809 rows x 1 columns]

In [ ]:
mest_enc.mapping

{'drinks': drinks
  1    0.189199
  2    0.148060
  3    0.170094
  4    0.195398
  5    0.179677
  6    0.139352
  7    0.172031
 -1    0.184674
 -2    0.184674
 dtype: float64}

## Weight of Evidence Encoder

In [102]:
from category_encoders import WOEEncoder

In [103]:
woe_enc=WOEEncoder()
woe_enc.fit_transform(df[col_cat],df['Class_enc'])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      0.029719
4      0.029719
6      0.029719
8      0.029719
10    -0.263874
...         ...
59940  0.029719
59943  0.070806
59944 -0.263874
59945  0.029719
59946  0.029719

[38809 rows x 1 columns]

In [104]:
woe_enc.mapping

{'drinks': drinks
  1    0.029719
  2   -0.263874
  3   -0.099262
  4    0.070806
  5   -0.030446
  6   -0.328956
  7   -0.069086
 -1    0.000000
 -2    0.000000
 dtype: float64}